File: component_example/executors_cpp/src/multi_threaded_executor.cpp
#include "rclcpp/rclcpp.hpp"
#include <chrono>
#include <thread>

using namespace std::chrono_literals;

class Node1 : public rclcpp::Node
{
public:
    Node1() : Node("node1")
    {
        cb_group1_ = this->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
        cb_group2_ = this->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
        this->timer1_ = this->create_wall_timer(
            1000ms, std::bind(&Node1::callbackTimer1, this), cb_group1_);
        this->timer2_ = this->create_wall_timer(
            1000ms, std::bind(&Node1::callbackTimer2, this), cb_group2_);
        this->timer3_ = this->create_wall_timer(
            1000ms, std::bind(&Node1::callbackTimer3, this), cb_group2_);
    }

private:

    void callbackTimer1()
    {
        std::this_thread::sleep_for(2000ms);
        RCLCPP_INFO(this->get_logger(), "cb 1");
    }

    void callbackTimer2()
    {
        std::this_thread::sleep_for(2000ms);
        RCLCPP_INFO(this->get_logger(), "cb 2");
    }

    void callbackTimer3()
    {
        std::this_thread::sleep_for(2000ms);
        RCLCPP_INFO(this->get_logger(), "cb 3");
    }

    rclcpp::TimerBase::SharedPtr timer1_;
    rclcpp::TimerBase::SharedPtr timer2_;
    rclcpp::TimerBase::SharedPtr timer3_;

    rclcpp::CallbackGroup::SharedPtr cb_group1_;
    rclcpp::CallbackGroup::SharedPtr cb_group2_;
};

class Node2 : public rclcpp::Node
{
public:
    Node2() : Node("node2")
    {
        cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
        this->timer4_ = this->create_wall_timer(
            1000ms, std::bind(&Node2::callbackTimer4, this), cb_group_);
        this->timer5_ = this->create_wall_timer(
            1000ms, std::bind(&Node2::callbackTimer5, this), cb_group_);
    }

private:

    void callbackTimer4()
    {
        std::this_thread::sleep_for(2000ms);
        RCLCPP_INFO(this->get_logger(), "cb 4");
    }

    void callbackTimer5()
    {
        std::this_thread::sleep_for(2000ms);
        RCLCPP_INFO(this->get_logger(), "cb 5");
    }

    rclcpp::TimerBase::SharedPtr timer4_;
    rclcpp::TimerBase::SharedPtr timer5_;

    rclcpp::CallbackGroup::SharedPtr cb_group_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node1 = std::make_shared<Node1>();
    auto node2 = std::make_shared<Node2>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node1);
    executor.add_node(node2);
    executor.spin();
    rclcpp::shutdown();
    return 0;
}

File: component_example/executors_py/executors_py/multi_threaded_executor.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup, MutuallyExclusiveCallbackGroup
import time


class Node1(Node):
    def __init__(self):
        super().__init__("node1")
        self.cb_group1_ = MutuallyExclusiveCallbackGroup()
        self.cb_group2_ = MutuallyExclusiveCallbackGroup()
        self.timer1_ = self.create_timer(
            1.0, self.callback_timer1, callback_group=self.cb_group1_)
        self.timer2_ = self.create_timer(
            1.0, self.callback_timer2, callback_group=self.cb_group2_)
        self.timer3_ = self.create_timer(
            1.0, self.callback_timer3, callback_group=self.cb_group2_)

    def callback_timer1(self):
        time.sleep(2.0)
        self.get_logger().info("cb 1")

    def callback_timer2(self):
        time.sleep(2.0)
        self.get_logger().info("cb 2")

    def callback_timer3(self):
        time.sleep(2.0)
        self.get_logger().info("cb 3")


class Node2(Node):
    def __init__(self):
        super().__init__("node2")
        self.cb_group_ = ReentrantCallbackGroup()
        self.timer4_ = self.create_timer(
            1.0, self.callback_timer4, callback_group=self.cb_group_)
        self.timer5_ = self.create_timer(
            1.0, self.callback_timer5, callback_group=self.cb_group_)

    def callback_timer4(self):
        time.sleep(2.0)
        self.get_logger().info("cb 4")

    def callback_timer5(self):
        time.sleep(2.0)
        self.get_logger().info("cb 5")


def main(args=None):
    rclpy.init(args=args)
    node1 = Node1()
    node2 = Node2()
    executor = MultiThreadedExecutor()
    executor.add_node(node1)
    executor.add_node(node2)
    executor.spin()
    rclpy.shutdown()


if __name__ == "__main__":
    main()